Objects/call.c

```diff
@@ -1298,59 +1295,80 @@ _PyStack_AsDict(PyObject *const *values, PyObject *kwnames)
 }
 
 
-int
-_PyStack_UnpackDict(PyObject *const *args, Py_ssize_t nargs, PyObject *kwargs,
-                    PyObject *const **p_stack, PyObject **p_kwnames)
+PyObject *const *
+_PyStack_DictAsVector(PyObject *const *args, Py_ssize_t nargs,
+                      PyObject *kwargs, PyObject **p_kwnames)
 {
-    PyObject **stack, **kwstack;
-    Py_ssize_t nkwargs;
-    Py_ssize_t pos, i;
-    PyObject *key, *value;
-    PyObject *kwnames;
-
     assert(nargs >= 0);
-    assert(kwargs == NULL || PyDict_CheckExact(kwargs));
+    assert(args != NULL || nargs == 0);
 
-    if (kwargs == NULL || (nkwargs = PyDict_GET_SIZE(kwargs)) == 0) {
-        *p_stack = args;
-        *p_kwnames = NULL;
-        return 0;
+    Py_ssize_t nkwargs = 0;
+    if (kwargs != NULL) {
+        assert(PyDict_Check(kwargs));
+        nkwargs = PyDict_GET_SIZE(kwargs);
     }
 
-    if ((size_t)nargs > PY_SSIZE_T_MAX / sizeof(stack[0]) - (size_t)nkwargs) {
-        PyErr_NoMemory();
-        return -1;
+    if (nkwargs == 0) {
+        *p_kwnames = NULL;
+        /* If there are no arguments, we don't want to return NULL because that
+         * would mean an error. Return an arbitrary non-NULL pointer. */
+        if (args == NULL) {
+            return args + 1;
+        }
+        else {
+            return args;
+        }
     }
 
-    stack = PyMem_Malloc((nargs + nkwargs) * sizeof(stack[0]));
-    if (stack == NULL) {
-        PyErr_NoMemory();
-        return -1;
+    /* Allocate one big tuple for everything, plus one sentinel NULL */
+    Py_ssize_t n = 2 * nkwargs + nargs + 1;
+    PyObject *kwtuple = PyTuple_New(n);
+    *p_kwnames = kwtuple;
+    if (kwtuple == NULL) {
+        return NULL;
     }
+    /* Truncate the tuple to the desired length */
+    Py_SIZE(kwtuple) = nkwargs;
```

Comment:
That doesn't look safe at all. Won't such a tuple join the wrong freelist in `tupledealloc`?